kore-rpc-types/src/Kore/Network/JsonRpc.hs

```diff
@@ -0,0 +1,175 @@
+{-# LANGUAGE TypeApplications #-}
+
+{- |
+Module      : Kore.Network.JSONRPC
+Description : JSON RPC server for kore-rpc
+Copyright   : (c) Runtime Verification, 2022
+License     : BSD-3-Clause
+-}
+module Kore.Network.JsonRpc (jsonRpcServer, JsonRpcHandler (..)) where
+
+import Control.Concurrent (forkIO, throwTo)
+import Control.Concurrent.STM.TChan (newTChan, readTChan, writeTChan)
+import Control.Exception (ErrorCall (..), Exception (fromException), SomeException, catch, mask, throw)
+import Control.Monad (forever)
+import Control.Monad.IO.Class (MonadIO (liftIO))
+import Control.Monad.Logger (MonadLoggerIO)
+import Control.Monad.Logger qualified as Log
+import Control.Monad.Reader (MonadReader (ask), ReaderT, runReaderT)
+import Data.Aeson (ToJSON, Value (Null))
+import Data.Aeson.Encode.Pretty as Json
+import Data.ByteString (ByteString)
+import Data.ByteString.Lazy.Char8 qualified as L8
+import Data.Conduit (ConduitT, Void, runConduit, (.|))
+import Data.Conduit.List qualified as CL
+import Data.Conduit.Network (ServerSettings, appSink, appSource, runGeneralTCPServer)
+import Data.Conduit.TMChan (closeTBMChan, sinkTBMChan, sourceTBMChan)
+import Data.Maybe (catMaybes)
+import Data.Text qualified as Text
+import Kore.JsonRpc.Types (FromRequestCancellable (isCancel), ReqException (..))
+import Network.JSONRPC hiding (encodeConduit, runJSONRPCT)
+import UnliftIO (MonadUnliftIO, atomically, wait, withAsync)
+
+-- Conduit to encode JSON to ByteString.
+encodeConduit :: (ToJSON j, Monad m) => Json.Config -> ConduitT j ByteString m ()
+encodeConduit encodeOpts = CL.mapM $ \m -> return . L8.toStrict $ encodePretty' encodeOpts m
+
+{- | Create JSON-RPC session around conduits from transport layer.  When
+ context exits session disappears.
+-}
+runJSONRPCT ::
+    (MonadLoggerIO m, MonadUnliftIO m) =>
+    -- | aeson-pretty Config
+    Json.Config ->
+    -- | JSON-RPC version
+    Ver ->
```

Comment:
Should we maybe just hard-wire the version to `V2`?
Only `initSession` and `decodeConduit` are using it, and  AFAICT using a different version (`V1`) would probably require a different request type anyway.